src/rules/spaceBeforeFunctionParenRule.ts

```diff
@@ -0,0 +1,173 @@
+import * as Lint from "../index";
+
+const ALWAYS_OR_NEVER = {
+  enum: ["always", "never"],
+  type: "string",
+};
+
+export class Rule extends Lint.Rules.AbstractRule {
+  public static metadata: Lint.IRuleMetadata = {
+    description: "Require or disallow a space before function parenthesis",
+    optionExamples: [`[true, {"anonymous": "always", "named": "never", "asyncArrow": "always"}]`],
+    options: {
+      properties: {
+        anonymous: ALWAYS_OR_NEVER,
+        asyncArrow: ALWAYS_OR_NEVER,
+        constructor: ALWAYS_OR_NEVER,
+        method: ALWAYS_OR_NEVER,
+        named: ALWAYS_OR_NEVER,
+      },
+      type: "object",
+    },
+    optionsDescription: Lint.Utils.dedent`
+      One argument which is an object which may contain the keys \`anonymous\`, \`named\`, and \`asyncArrow\`
+      These should be set to either \`"always"\` or \`"never"\`.
+
+      * \`"anonymous"\` checks before the opening paren in anonymous functions
+      * \`"named"\` checks before the opening paren in named functions
+      * \`"asyncArrow"\` checks before the opening paren in async arrow functions,
+      * \`"method"\` checks before the opening paren in class methods
+      *\`"constructor"\` checks before the opening paren in class constructors
+    `,
+    ruleName: "space-before-function-paren",
+    type: "style",
+    typescriptOnly: false,
+  };
+  public static INVALID_WHITESPACE_ERROR = "Should be no space before paren";
+  public static MISSING_WHITESPACE_ERROR = "Should be a space before paren";
+
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new FunctionWalker(sourceFile, this.getOptions()));
+  }
+}
+
+class FunctionWalker extends Lint.RuleWalker {
+  private scanner: ts.Scanner;
+
+  constructor(sourceFile: ts.SourceFile, options: Lint.IOptions) {
+    super(sourceFile, options);
+    this.scanner = ts.createScanner(ts.ScriptTarget.ES5, false, ts.LanguageVariant.Standard, sourceFile.text);
+  }
+
+  protected visitArrowFunction(node: ts.FunctionLikeDeclaration): void {
+    this._visitFunction(node);
+    super.visitArrowFunction(node);
+  }
+
+  public visitConstructorDeclaration(node: ts.ConstructorDeclaration): void {
+    this._visitFunction(node);
+    super.visitConstructorDeclaration(node);
+  }
+
+  protected visitFunctionDeclaration(node: ts.FunctionDeclaration): void {
+    this._visitFunction(node);
+    super.visitFunctionDeclaration(node);
+  }
+
+  protected visitFunctionExpression(node: ts.FunctionExpression): void {
+    this._visitFunction(node);
+    super.visitFunctionExpression(node);
+  }
+
+  public visitMethodDeclaration(node: ts.MethodDeclaration): void {
+    this._visitFunction(node);
+    super.visitMethodDeclaration(node);
+  }
+
+  public visitMethodSignature(node: ts.SignatureDeclaration): void {
+    this._visitFunction(node);
+    super.visitMethodSignature(node);
+  }
+
+  private _visitFunction(node: ts.Node): void {
```

Comment:
we don't use the underscore naming convention -- the `private` modifier is enough to denote encapsulation.